File: python/bKap5_6_InSample.py
"""
bKap5_6_InSample.py
Python port of bKap5_6_InSample.m.

PLS dimensionality reduction of the Liedtke macro panel (in-sample). The macro
predictors are reduced to a few PLS components (which, unlike PCA, are extracted
to explain the inflation target) and inflation is regressed on those components.

Run from the repo root:  python python/bKap5_6_InSample.py
"""

import os
import numpy as np
import pandas as pd
import scipy.io as sio

import models

# =========================================================================
#  Settings  (mirror the MATLAB header)
# =========================================================================
DATA_PATH    = "./DATA/Liedtke/US/aggregated.csv"
RESULTS_PATH = "./RESULTS/GWZ/"

iTimeLag    = 1     # additional predictive lag (predictors already lag-aligned)
iNumComp    = 3     # number of PLS components to retain
iTransformX = 2     # 0 = none, 1 = centre, 2 = z-standardise

# =========================================================================
#  Load data
# =========================================================================
tData = pd.read_csv(DATA_PATH)
cAllNames = list(tData.columns)
dtDates = pd.to_datetime(tData.iloc[:, 0])
mX = tData.iloc[:, 2:].to_numpy(float)
vY = tData.iloc[:, 1].to_numpy(float)
cXnamesM = cAllNames[2:]

iNumObs, iNumPredictors = mX.shape

# Lag the predictors so they are known one period ahead of the target
mXlag = np.full_like(mX, np.nan)
mXlag[iTimeLag:, :] = mX[:-iTimeLag, :]

# Remove missing values row-wise (PLS does not allow missing values)
lIsNaN = np.isnan(vY) | np.isnan(mXlag).any(axis=1)
keep = ~lIsNaN
vY = vY[keep]
mXlag = mXlag[keep, :]
dtDates = dtDates[keep].reset_index(drop=True)

iNumObs = len(vY)
iNumComp = min(iNumComp, iNumPredictors)

# =========================================================================
#  PLS dimensionality reduction
# =========================================================================
rModel = models.est_pls(vY, mXlag, num_comp=iNumComp, transform_x=iTransformX)
mScores = rModel["scores"]

# =========================================================================
#  In-sample analysis: regress inflation on the retained PLS comps (with const)
# =========================================================================
reg = models.ols_regress(vY, mScores)
vBeta = reg["beta"]
vBetaT = reg["t"]
dR2 = reg["R2"]
vYhat = reg["yhat"]

# =========================================================================
#  Report
# =========================================================================
print(f"In-sample PLS regression: inflation ~ {iNumComp} PLS components")
print(f"  Variance explained by {iNumComp} comps : "
      f"X {rModel['expl_var'][0, -1] * 100:6.2f}% | "
      f"Y {rModel['expl_var'][1, -1] * 100:6.2f}%")
print(f"  Intercept     : {vBeta[0]:8.4f} (t = {vBetaT[0]:6.2f})")
for k in range(iNumComp):
    print(f"  Comp{k + 1:<2d} coef   : {vBeta[k + 1]:8.4f} (t = {vBetaT[k + 1]:6.2f})")
print(f"  R2            : {dR2:8.4f}")

# =========================================================================
#  Save results (Python copy; does not overwrite the MATLAB .mat)
# =========================================================================
os.makedirs(RESULTS_PATH, exist_ok=True)
sio.savemat(os.path.join(RESULTS_PATH, "InSamplePLSResults_py.mat"), {
    "vBeta": vBeta.reshape(-1, 1),
    "vBetaT": vBetaT.reshape(-1, 1),
    "dR2": dR2,
    "vYhat": vYhat.reshape(-1, 1),
    "rModel": {
        "vMeanX": rModel["mean_x"], "vStdX": rModel["std_x"],
        "vMeanY": rModel["mean_y"], "vStdY": rModel["std_y"],
        "iNumComp": rModel["num_comp"], "mWeightsX": rModel["weights_x"],
        "mScores": rModel["scores"], "mExplVar": rModel["expl_var"],
    },
})

File: bKap5_6_InSample.m
% Script for applying PLS dimensionality reduction to the Liedtke macro panel
% (in-sample). The macro predictors are reduced to a few PLS components (which,
% unlike PCA, are extracted to explain the inflation target) and inflation is
% regressed on those components.

% Clear console
clear; clc; close all;

% Set paths
sOldPath = path;
sDataPath = './DATA/Liedtke/US/';
sResultsPath = './RESULTS/GWZ/';
addpath('./Utils/');

%% Settings
iTimeLag    = 1;     % additional predictive lag (predictors are already
                     % reporting-lag aligned in DATA/Liedtke/aggregate.py)
iNumComp    = 3;     % number of PLS components to retain
iTransformX = 2;     % preprocessing: 0 = none, 1 = centre, 2 = z-standardise.
                     % The predictors live on very different scales (rate diffs
                     % vs percentage changes), so standardisation is used.

%% Load data
% Macro panel produced by DATA/Liedtke/aggregate.py.
% CSV layout: col 1 = observation_date, col 2 = target (inflation),
%             col 3..end = predictors (already lag-aligned in Python).
tData     = readtable([sDataPath, 'aggregated.csv']);
cAllNames = tData.Properties.VariableNames;

% Parse dates (first column)
dtDates = tData{:,1};
if ~isdatetime(dtDates)
    dtDates = datetime(dtDates);
end

% Target (col 2) and predictors (col 3..end)
mX          = tData{:,3:end};          % macro predictors
vY          = tData{:,2};              % inflation target
cXnamesM    = cAllNames(3:end)';       % predictor names as an Nx1 cell

%% Data preprocessing
% Determine dimensions
[iNumObs, iNumAssets]       = size(vY);
[iNumObsP, iNumPredictors]  = size(mX);

% Lag the predictors so that they are known one period ahead of the target
mXlag = [NaN(iTimeLag, iNumPredictors); mX(1:end-iTimeLag,:)];

% Remove missing values row-wise (PLS does not allow missing values)
lIsNaN          = isnan(vY) | any(isnan(mXlag),2);
vY(lIsNaN)      = [];
mXlag(lIsNaN,:) = [];
dtDates(lIsNaN) = [];

% Update number of observations after NaN removal
iNumObs  = size(vY,1);
iNumComp = min(iNumComp, iNumPredictors);

%% PLS dimensionality reduction
% Estimate PLS keeping iNumComp components
rModel  = fEstPLS(vY, mXlag, "iNumComp", iNumComp, "iTransformX", iTransformX);
mScores = rModel.mScores;

%% In-Sample analysis
% Regression with intercept using the retained PLS components
rResults = regstats2(vY, mScores, 'linear', {'tstat','rsquare'});
vBeta    = rResults.tstat.beta;   % coefficient estimates
vBetaT   = rResults.tstat.t;      % t-Statistics
dR2      = rResults.rsquare;      % R2

% In-sample prediction
vYhat = [ones(iNumObs,1), mScores] * vBeta;

% === Report
fprintf('In-sample PLS regression: inflation ~ %d PLS components\n', iNumComp);
fprintf('  Variance explained by %d comps : X %6.2f%% | Y %6.2f%%\n', iNumComp, ...
    rModel.mExplVar(1,end)*100, rModel.mExplVar(2,end)*100);
fprintf('  Intercept     : %8.4f (t = %6.2f)\n', vBeta(1), vBetaT(1));
for k = 1:iNumComp
    fprintf('  Comp%-2d coef   : %8.4f (t = %6.2f)\n', k, vBeta(k+1), vBetaT(k+1));
end
fprintf('  R2            : %8.4f\n', dR2);

% === Save results
sFilename = [sResultsPath, 'InSamplePLSResults.mat'];
save(sFilename, 'vBeta', 'vBetaT', 'dR2', 'vYhat', 'rModel');

% Restore path
path(sOldPath);
